openverse_catalog/dags/providers/provider_api_scripts/nypl.py

```diff
@@ -1,211 +1,248 @@
-        title = titleinfo[0].get("title", {}).get("$")
-    return title
-
-
-def _get_creators(creatorinfo):
-    if type(creatorinfo) == list:
+        captures = item_details.get("sibling_captures", {}).get("capture")
+        if not captures:
+            return None
+        if not isinstance(captures, list):
+            captures = [captures]
+        images = []
+        for capture in captures:
+            image_id = capture.get("imageID", {}).get("$")
+            if image_id is None:
+                continue
+
+            image_link = capture.get("imageLinks", {}).get("imageLink", [])
+            image_url, filetype = self._get_image_data(image_link)
+            if not image_url:
+                continue
+
+            foreign_landing_url = capture.get("itemLink", {}).get("$")
+            license_url = capture.get("rightsStatementURI", {}).get("$")
+            if not foreign_landing_url or license_url is None:
+                continue
+
+            image_data = {
+                "foreign_identifier": image_id,
+                "foreign_landing_url": foreign_landing_url,
+                "image_url": image_url,
+                "license_info": get_license_info(license_url=license_url),
+                "title": title,
+                "creator": creator,
+                "filetype": filetype,
+                "category": category,
+                "meta_data": metadata,
+            }
+            images.append(image_data)
+        return images or None
+
+    @staticmethod
+    def _get_filetype(description: str):
+        """
+        Extracts the filetype from a description string like:
+        "Cropped .jpeg (1600 pixels on the long side)"
+        :param description: the description string
+        :return:  jpeg | gif
+        """
+        if match := re.search(FILETYPE_PATTERN, description):
+            return match.group(1)
+        return None
+
+    @staticmethod
+    def _get_image_data(images):
+        """
+        Gets a list of dictionaries of the following shape:
+        {
+          "$": "http://images.nypl.org/index.php?id=56738467&t=q&download=1
+        &suffix=29eed1f0-3d50-0134-c4c7-00505686a51c.001",
+          "description": "Cropped .jpeg (1600 pixels on the long side)"
+        }
+        Selects the largest image based on the image URL's `t` query parameter
+        and IMAGE_URL_DIMENSIONS.
+        """
+        # Create a dict with the NyplDataIngester.image_url_dimensions as keys,
+        # and image data as value.
+        image_types = {
+            parse_qs(urlparse(img["$"]).query)["t"][0]: i
+            for i, img in enumerate(images)
+        }
+        if not image_types:
+            return None, None
+
+        # Select the dict with the largest image
+        preferred_image_index = next(
+            image_types[dimension]
+            for dimension in NyplDataIngester.image_url_dimensions
+            if dimension in image_types
+        )
```

Comment:
If none of the possible dimensions are available, this will raise a `StopIteration` exception:
```python
[ins] In [2]: next(x for x in [])
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ <ipython-input-2-dbd4591bda43>:1 in <module>                                                     │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
StopIteration
```

Does that have a possibility of occurring? Should we construct this as a list then see if it has any length rather than using `next`?

Reply:
Thank you, I didn't know this! I've refactored the code here.